src/librc/librc-misc.c

```diff
@@ -447,3 +450,51 @@ rc_conf_value(const char *setting)
 
 	return rc_config_value(rc_conf, setting);
 }
+
+bool rc_export_variable(const char *service, const char *name, const char *value) {
+	int dirfd = openat(rc_dirfd(RC_DIR_ENVIRONMENT), service, O_RDONLY | O_DIRECTORY);
+	FILE *envfile;
+
+	if (dirfd == -1)
+		return false;
+
+	if (!value) {
+		unlinkat(dirfd, name, 0);
+		return true;
+	}
+
+	envfile = do_fopenat(dirfd, name, O_WRONLY | O_TRUNC | O_CREAT);
+	close(dirfd);
+
+	if (!envfile)
+		return false;
+
+	/* in case of a race for the same variable, for the same service
+	 * at the same time, let's just use the first that locked. */
+	if (flock(fileno(envfile), LOCK_EX) == -1) {
+		fclose(envfile);
+		return false;
+	}
+
+	fputs(value, envfile);
+	return fclose(envfile) == 0;
+}
+
+void rc_import_variables(const char *service) {
+	DIR *envdir = do_opendirat(rc_dirfd(RC_DIR_ENVIRONMENT), service ? service : "rc");
+	struct dirent *d;
+	if (!envdir)
+		return;
+
+	while ((d = readdir(envdir))) {
+		char *value;
+		size_t size;
+
+		if (!do_getfileat(rc_dirfd(RC_DIR_ENVIRONMENT), d->d_name, &value, &size))
+			continue;
```

Comment:
Same here. Silently ignoring actual io failure can make it hard to debug when it actually happens.